backend/src/test/java/com/google/backend/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,142 @@
+package com.google.backend.servlets;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.Mockito.*;
+import org.mockito.MockitoAnnotations;
+
+/** Contain tests for methods in {@link LoginServlet} class. */
+@RunWith(JUnit4.class)
+public class LoginServletTest {
+  @Mock HttpServletRequest request;
+  @Mock HttpServletResponse response;
+
+  private static final LoginServlet loginServlet = new LoginServlet();
+  private final LocalServiceTestHelper helper = new LocalServiceTestHelper(
+    new LocalDatastoreServiceTestConfig(), new LocalUserServiceTestConfig());
+
+  private static final String CONTENT_TYPE = "text/html;";
+  private static final String TEST_EMAIL = "test@example.com";
+  private static final String REDIRECT_URL = "/";
+  private static final String LOGIN_STATUS = "logged in";
+  private static final String LOGOUT_STATUS = "stranger";
+  private static final String ENTITY_KIND = "User";
+  private static final int MAX_QUERIES = 1000;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGet_GetsCorrectResponseLoggedIn() throws IOException, ServletException {
+    helper.setEnvIsLoggedIn(true).setEnvEmail(TEST_EMAIL).setEnvAuthDomain("localhost");
+    UserService userService = UserServiceFactory.getUserService();
+    String logoutUrl = userService.createLogoutURL(REDIRECT_URL);
+    String loginResponse = LOGIN_STATUS + "\n" + logoutUrl;
+    
+    StringWriter stringWriter = new StringWriter();
```

Comment:
I see that you have the following lines:
```
StringWriter stringWriter = new StringWriter();
PrintWriter writer = new PrintWriter(stringWriter);
when(response.getWriter()).thenReturn(writer);
```
in every one of your tests in this file. Since `stringWriter` and `writer` are declared the same in every one of your test functions, you can just have all the functions share the same `stringWriter` and `writer`. 

i think you can do something like this:
```
public class LoginServletTest {
  // other private var declaration
  private StringWriter stringWriter = new StringWriter();

  @Before
  public void setUp() throws Exception {
    MockitoAnnotations.initMocks(this);
    helper.setUp();
    when(response.getWriter()).thenReturn(new PrintWriter(stringWriter));
  }

  // ... rest of code
}
```
`setUp()` gets run before every single one of your tests for junit, same for `tearDown()`, that's why you can move the shared setup for all your tests there to reduce redundancy :) 

lmk if this doesn't work or you prefer another way!

Reply:
yeah... i tried it like that but the stringwriter didn't get cleared for the next test, so that's why I figured I had to redo it, idk why it didn't get cleared :/ 
is there a way to clear it?

Reply:
do you mean that moving the `stringWriter` to either setup or as `private` variable makes the rest of the test not pass?

i tried smthing like this in your test code:
![image](https://user-images.githubusercontent.com/34081368/89759070-352c3900-da9e-11ea-9bb3-7ec9c9f45a57.png)
and it seems to still pass all your tests...

but I could be wrong :') 

how did you know that `stringWriter` didnt get cleared correctly?

Reply:
it didn't pass a test b/c the response included the output from the previous test... I will try it again

Reply:
sigh. idk why it didn't work before lol but it works now. will update the code